File: 3. Behavioral Design Patterns/State/Go/main.go
package main

import "fmt"

// State Pattern Example in Go
// TrafficLight cycles through Red, Yellow, Green states

type TrafficLightState interface{ Next(*TrafficLight); GetColor() string }

type TrafficLight struct{ state TrafficLightState }
func NewTrafficLight() *TrafficLight { return &TrafficLight{state: &RedState{}} }
func (t *TrafficLight) SetState(s TrafficLightState) { t.state = s }
func (t *TrafficLight) Next() { t.state.Next(t) }
func (t *TrafficLight) GetColor() string { return t.state.GetColor() }

type RedState struct{}
func (r *RedState) Next(t *TrafficLight) { t.SetState(&GreenState{}) }
func (r *RedState) GetColor() string { return "Red" }

type YellowState struct{}
func (y *YellowState) Next(t *TrafficLight) { t.SetState(&RedState{}) }
func (y *YellowState) GetColor() string { return "Yellow" }

type GreenState struct{}
func (g *GreenState) Next(t *TrafficLight) { t.SetState(&YellowState{}) }
func (g *GreenState) GetColor() string { return "Green" }

func main(){
	light := NewTrafficLight()
	for i := 0; i < 6; i++ {
		fmt.Printf("Light: %s\n", light.GetColor())
		light.Next()
	}
}

File: 3. Behavioral Design Patterns/State/Rust/main.rs
// State Pattern Example in Rust
// TrafficLight cycles through Red, Yellow, Green states

trait TrafficLightState {
    fn next(self: Box<Self>) -> Box<dyn TrafficLightState>;
    fn get_color(&self) -> &str;
}

struct TrafficLight { state: Box<dyn TrafficLightState> }
impl TrafficLight {
    fn new() -> Self { Self { state: Box::new(RedState) } }
    fn next(&mut self) { let old_state = std::mem::replace(&mut self.state, Box::new(RedState)); self.state = old_state.next(); }
    fn get_color(&self) -> &str { self.state.get_color() }
}

struct RedState;
impl TrafficLightState for RedState { fn next(self: Box<Self>) -> Box<dyn TrafficLightState> { Box::new(GreenState) } fn get_color(&self) -> &str { "Red" } }

struct YellowState;
impl TrafficLightState for YellowState { fn next(self: Box<Self>) -> Box<dyn TrafficLightState> { Box::new(RedState) } fn get_color(&self) -> &str { "Yellow" } }

struct GreenState;
impl TrafficLightState for GreenState { fn next(self: Box<Self>) -> Box<dyn TrafficLightState> { Box::new(YellowState) } fn get_color(&self) -> &str { "Green" } }

fn main(){
    let mut light = TrafficLight::new();
    for _ in 0..6 { println!("Light: {}", light.get_color()); light.next(); }
}
